File: frontend/app/models/relationship.js
import Ember from 'ember';
import DS from 'ember-data';

const { computed } = Ember;

export default DS.Model.extend({

  required: DS.attr('boolean'),

  cardinality: DS.attr('string'),

  entity: DS.belongsTo('entity', { inverse: 'relationships' }),
  dependentEntity: DS.belongsTo('entity', { inverse: 'sourceRelationships' }),

  dependentEntityLoaded: computed('dependentEntity.name', function() {
    let dependentEntity = this.belongsTo('dependentEntity').value();
    return dependentEntity !== null;
  })

});

File: app/models/relationship.rb
class Relationship < ApplicationRecord

  belongs_to :entity
  belongs_to :dependent_entity, class_name: 'Entity', foreign_key: 'dependent_entity_id', inverse_of: :source_relationships

  scope :required, -> { where required: true }

  def hyphenized_name
    relationship_name = dependent_entity.name
    if is_plural?
      relationship_name = relationship_name.pluralize
    end

    relationship_name.downcase.strip.gsub(' ', '-').gsub(/[^\w-]/, '')
  end

  def is_plural?
    cardinality === 'HAS_MANY'
  end

end
